### DISCUSSION TITLE
Planning: Upcoming changes in IKOS 3.3

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [NASA-SW-VnV/ikos#258] [General]
IKOS 3.3 is planned to be released on March 28, 2024.

--

This is just an open discussion to propose issues to address in IKOS 3.3. If there's a bug that is affecting you, one that you know the solution for, or a feature you'd like to contribute, feel free to comment below.

EDIT: Roughly, the plan is as follows:

- Dec 28 - Jan 10: Planning and selecting issues.
- Jan 11 - Mar 7: Development

  -- *Mar 8: Request to release*

- Mar 8 - Mar 14: First release candidate, testing
- Mar 15 - Mar 21: Second release candidate / draft of final release, testing
- Mar 25 - Mar 28: Draft release announcement

  -- *Mar 28: Final release*

**Comment by ivanperez-keera** [comment]
- Issue #221 looks like low-hanging fruit.
- Issue #197 has a proposed solution already.

**Reply by ivanperez-keera** [reply]
Regarding issue #197, I'm not yet equipped to evaluate it in depth; all I can do at this point is check whether the bug does not manifest after applying the proposed solution. I don't know how else it could affect IKOS.

**Reply by ivanperez-keera** [reply]
I would love to see https://github.com/NASA-SW-VnV/ikos/issues/194 fixed, but when I tried to fix it myself, I could not locate the actual source of the bug. I can reproduce the issue, though.

**Reply by ivanperez-keera** [reply]
Installation has been an issue. We have some dockerfiles but they are outdated. Someone (post 3.2 release) pointed out that there is an installation issue in Fedora. I'd like to have CI jobs and test IKOS in every single platform we can possibly support.

We may have to execute those manually prior to release if they take too long to run or if we are limited in CI minutes in some way.

Also, our current CI job only compiles IKOS, but I'd like to see IKOS being tested, and not only the IKOS tool, but other tools we are providing.

**Reply by ivanperez-keera** [reply]
I also do not understand why the python scripts live under `/analyzer/`. To me, IKOS is two packages: a C++ library, and a set of python tools.

These two should live in separate paths in the repository, because, although related, they are not the same, there's a clear direction in dependencies.

If we were to make these available in Debian/Ubuntu, which I hope we do, I'd expect there two be too packages: `libikos-dev` (the library) and `ikos` (the tools).

If we separate the two in the repo, I think it'll be easier to publish packages, and even installation will become easier for us.

**Reply by ivanperez-keera** [reply]
More things I'd like to address at some point:

- IKOS produces a lot of false positives. The default settings should produce fewer false positives, otherwise people don't use the tools because of the signal to noise ratio.

- IKOS is advertised as a C/C++ static analyzer, but C++ support is preliminary and most of the times, when someone has an issue, we just say "sorry, not supported". I think it'd be better to be straightforward and say it's a C static analyzer, and then somewhere else say that we have preliminary support for C++.

- I'd like to collect a list of orgs using IKOS, and evaluate the user experience more deeply.

- IKOS is advertised as "Static analyzer for C/C++ based on the theory of Abstract Interpretation." I think we lose half the crowd there because they don't know what the "theory of Abstract Interpretation" is. Let's just advertise it as a "Static analyzer for C" with some additional qualifiers that convey to the person not versed in abstract interpretation what the distinguishing characteristics are.

**Reply by ivanperez-keera** [reply]
And to help drive this discussion, here are some goals for IKOS in the longer term future:

- Increase stability.

- Increase accuracy.

- Increase adoption.

- Increase active participation from the open source community.

**Reply by ivanperez-keera** [reply]
I don't know exactly what it would entail, but we soon may have to start considering supporting LLVM > 14.

**Reply by ivanperez-keera** [reply]
I'd like to see this issue investigated and hopefully solved: https://github.com/NASA-SW-VnV/homebrew-core/issues/11

**Reply by ivanperez-keera** [reply]
Similarly to 197, for https://github.com/NASA-SW-VnV/ikos/issues/205 we have a candidate solution, making it a perfect candidate to be included in the next milestone. I'm not yet equipped to evaluate it in depth; all I can do at this point is check whether the bug does not manifest after applying the proposed solution.

**Comment by ivanperez-keera** [comment]
@asensi has indicated that we may not have to bump the copyright years like we did for IKOS 3.2. The following comments offer more details:

- https://github.com/NASA-SW-VnV/ikos/discussions/257#discussioncomment-7980028
- https://github.com/NASA-SW-VnV/ikos/discussions/257#discussioncomment-7980047

**Comment by ivanperez-keera** [comment]
An upcoming change that we may consider for 3.3 is the change of license, which has been requested. This will allow IKOS to also be included in Debian's `main` repositories (right now we can only be included in `non-free`, which a lot of people do not enable).

We cannot really plan it for 3.3 as such, since it doesn't really depend on us and it can take an arbitrary amount of time. But we'll add it to the release if possible.

**Reply by asensi** [reply]
It seems the best that can be done at this time. Thanks!

**Comment by ivanperez-keera** [comment]
I started on a new project in Jan and could not finish as much as I had originally hoped to do. However, we still have a release and we can move some of these low-hanging issues to the next release.
